windows/src/engine/inst-cpp/insthelper.cpp

```diff
@@ -0,0 +1,211 @@
+#include "insthelper.h"
+#include "pch.h"
+#include <AccCtrl.h>
+#include <ShlObj_core.h>
+#include <iostream>
+#include <knownfolders.h>
+#include <limits.h>
+#include <msiquery.h>
+#include <security.h>
+#include <utility>
+#include <tchar.h>
+#include <string>
+#include <AclAPI.h>
+#include <Windows.h>
+#include <Msi.h>
+#include <initguid.h>
+#include <comdef.h>
+#include <msctf.h>
+#include <atlbase.h>
+#include<sstream>
+#include<iomanip>
+
+const LPCTSTR SFolderKeymanRoot = TEXT("\\Keyman");
+
+extern "C" __declspec(dllexport) unsigned int EnginePostInstall(MSIHANDLE hInstall) {
+  HANDLE hFile;
+
+  // Find %appdata% path
+  PWSTR path = new wchar_t[MAX_PATH];
+
+  if (SUCCEEDED(SHGetKnownFolderPath(FOLDERID_ProgramData, 0, NULL, &path))) {
+    wcscat_s(path, MAX_PATH, SFolderKeymanRoot);
+
+    // Create directory if it does not exist
+    if (GetFileAttributes(path) == INVALID_FILE_ATTRIBUTES) {
+      if (!CreateDirectory(path, NULL)) {
+        DWORD errorCode = GetLastError();
+        std::wstring error =
+            L"Keyman Engine failed to set permissions on shared data in CreateDir: " + std::to_wstring(errorCode);
+        MsiSetProperty(hInstall, TEXT("EnginePostInstall_Error"), error.c_str());
+        return errorCode;
```

Comment:
Can we turn this pattern into a helper function?

Reply:
Yes, this pattern is extracted as HandleError().